modules/apps/portal-url-builder/portal-url-builder-impl/src/main/java/com/liferay/portal/url/builder/internal/PortletDependencyAbsolutePortalURLBuilderImpl.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.url.builder.internal;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.model.portlet.PortletDependency;
+import com.liferay.portal.kernel.servlet.PortalWebResourceConstants;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.url.builder.PortletDependencyAbsolutePortalURLBuilder;
+import com.liferay.portal.url.builder.internal.util.URLUtil;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+public class PortletDependencyAbsolutePortalURLBuilderImpl
+	implements PortletDependencyAbsolutePortalURLBuilder {
+
+	public PortletDependencyAbsolutePortalURLBuilderImpl(
+		String cdnHost, String cssURN, String javaScriptURN, String pathProxy,
+		PortletDependency portletDependency) {
+
+		_cdnHost = cdnHost;
+		_cssURN = cssURN;
+		_javaScriptURN = javaScriptURN;
+		_pathProxy = pathProxy;
+		_portletDependency = portletDependency;
+	}
+
+	@Override
+	public String build() {
+		StringBundler relativeURLSB = new StringBundler(7);
+
+		boolean ignoreCDNHost = _ignoreCDNHost;
+		boolean ignorePathProxy = false;
+
+		if (PortletDependency.Type.CSS == _portletDependency.getType()) {
+			String resourcePath = _cssURN;
+
+			if (Validator.isNull(resourcePath)) {
+				resourcePath = PortalWebResourceConstants.RESOURCE_TYPE_CSS;
+			}
+			else {
+				ignoreCDNHost = true;
+				ignorePathProxy = true;
+			}
+
+			relativeURLSB.append(resourcePath);
+		}
+		else if (PortletDependency.Type.JAVASCRIPT ==
+					_portletDependency.getType()) {
+
+			String resourcePath = _javaScriptURN;
+
+			if (Validator.isNull(resourcePath)) {
+				resourcePath = PortalWebResourceConstants.RESOURCE_TYPE_JS;
+			}
+			else {
+				ignoreCDNHost = true;
+				ignorePathProxy = true;
+			}
+
+			relativeURLSB.append(resourcePath);
+		}
```

Comment:
What about image resource types? Do we need to add that as well?

Reply:
My plan is to add new resources as we need them. Since there are tons of manually composed URLs in the portal and we don't now the requirements to construct each one of them, we cannot design this API upfront :-(